### PR TITLE
specify boundary conditions as a tuple

### PR DESCRIPTION
**Pull request by simonbyrne**
### Description

The main changes are:
1. A `BalanceLaw` defines an additional function `boundary_conditions(::BL)` which should return a tuple of boundary conditions (currently these can be anything, but we may change that in future).
2. Instead of passing the integer `bctype` to the boundary condition functions, we now look up the corresponding element of the tuple and pass that (this matches how Atmos and Ocean were doing it internally, so we can remove that code).
3. I've moved the boundary condition argument to be the 2nd one, to make it easier to do dispatch.

- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [x] Documentation has been added/updated OR N/A.

### CODE REVIEW

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try-

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by glwagner** [comment]
What did the interface look like before this PR, and what is the interface that this PR implements?

**Comment by simonbyrne** [comment]
The main changes are:
1. A `BalanceLaw` defines an additional function `boundary_conditions(::BL)` which should return a tuple of boundary conditions (currently these can be anything, but we may change that in future).
2. Instead of passing the integer `bctype` to the boundary condition functions, we now look up the corresponding element of the tuple and pass that (this matches how Atmos and Ocean were doing it internally, so we can remove that code).
3. I've moved the boundary condition argument to be the 2nd one, to make it easier to do dispatch.

**Comment by simonbyrne** [comment]
bors try

File: src/BalanceLaws/boundaryconditions.jl [hunk 1]
```diff
@@ -0,0 +1,3 @@
+function boundary_conditions end
```
Comment (charleskawczynski) [inline src/BalanceLaws/boundaryconditions.jl:1]: Can we add some docs to this? It would be helpful to also know what this does in edge cases like `boundary_conditions(::BalanceLaw) = ()` and `boundary_conditions(::BalanceLaw) = (nothing,)`

Comment (charleskawczynski) [reply src/BalanceLaws/boundaryconditions.jl:1]: Maybe add an entry in `docs/src/APIs/BalanceLaws/BalanceLaws.md`, too.

**Review by charleskawczynski — COMMENTED** [review]
Overall, looks great! I'm happy to see the generated functions move out of the physics kernels, since that will likely reappear for every balance law. I think we just need to add some docs for the new BalanceLaw pieces.

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

File: src/Numerics/DGMethods/DGModel_kernels.jl [hunk 2]
```diff
@@ -900,74 +863,80 @@ fluxes, respectively.
...
+                        local_state_prognostic_bottom1,
+                    ),
+                    Vars{vars_state(balance_law, GradientFlux(), FT)}(
+                        local_state_gradient_flux_bottom1,
+                    ),
+                    Vars{vars_state(balance_law, Auxiliary(), FT)}(
+                        local_state_auxiliary_bottom1,
+                    ),
+                )
+            end d -> throw(BoundsError(bcs, bctag))
```
Comment (blallen) [inline src/Numerics/DGMethods/DGModel_kernels.jl:939]: @simonbyrne why did you have to unroll the tuple in the kernel versus in its own function like before? Curious as to why this solved the problems you were having

Comment (simonbyrne) [reply src/Numerics/DGMethods/DGModel_kernels.jl:939]: the next step after this is to be able to add an extra "boundary auxiliary state", which will depend on the boundary condition. For that to work, we would need to unroll it in the kernel.

**Review by blallen — APPROVED** [review]
Looks all good to me

File: src/Numerics/DGMethods/DGModel_kernels.jl (hunk 2 above)
Comment (blallen) [reply src/Numerics/DGMethods/DGModel_kernels.jl:939]: ah got it

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try-

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

File: test/Atmos/EDMF/edmf_kernels.jl [hunk 3]
```diff
@@ -742,24 +768,22 @@ function turbconv_normal_boundary_flux_second_order!(
...
+    # @unroll_map(N_up) do i
+    #     up_flx[i].ρaw = -n⁻ * FT(0)
+    #     up_flx[i].ρa = -n⁻ * FT(0)
+    #     up_flx[i].ρaθ_liq = -n⁻ * FT(0)
+    #     up_flx[i].ρaq_tot = -n⁻ * FT(0)
+    # end
+    # en_flx.∇tke = -n⁻ * FT(0)
+    # en_flx.∇e_int_cv = -n⁻ * FT(0)
+    # en_flx.∇q_tot_cv = -n⁻ * FT(0)
+    # en_flx.∇e_int_q_tot_cv = -n⁻ * FT(0)
```
Comment (simonbyrne) [inline test/Atmos/EDMF/edmf_kernels.jl:787]: @charleskawczynski This code didn't work (previously it wasn't being called, as the top boundary condition didn't use it).

**Comment by simonbyrne** [comment]
@blallen @christophernhill I'm getting failures on one of the ocean tests that I can't isolate:
https://buildkite.com/clima/climatemachine-ci/builds/680#41632c48-75e3-41a5-aa86-212493f38317

Any ideas?

File: test/Atmos/EDMF/edmf_kernels.jl (hunk 3 above)
Comment (charleskawczynski) [reply test/Atmos/EDMF/edmf_kernels.jl:787]: Ah, good to know, thank you. Feel free to remove it for now, I'll open an issue about adding top BCs back in. cc @yairchn

**Comment by blallen** [comment]
> @blallen @christophernhill I'm getting failures on one of the ocean tests that I can't isolate:
> https://buildkite.com/clima/climatemachine-ci/builds/680#41632c48-75e3-41a5-aa86-212493f38317
> 
> Any ideas?

Looks like the precision for the mean of `wz0` is slightly lower in this version: https://buildkite.com/clima/climatemachine-ci/builds/680#41632c48-75e3-41a5-aa86-212493f38317/276-530. The precision in the code is set to [11 decimal places](https://github.com/CliMA/ClimateMachine.jl/blob/6c17078487b84c7fca88c6801e49c2f04eccb4d2/test/Ocean/refvals/test_ocean_gyre_refvals.jl#L9), and it looks like we need to drop it to 9 or 10 decimal places.

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

File: test/Numerics/DGMethods/advection_diffusion/pseudo1D_heat_eqn.jl [hunk 4]
```diff
@@ -54,7 +54,12 @@ Dirichlet_data!(P::HeatEqn, x...) = initial_condition!(P, x...)
 
 function normal_boundary_flux_second_order!(
     ::CentralNumericalFluxSecondOrder,
+<<<<<<< HEAD
+    bctype,
+    ::AdvectionDiffusion{dim, HeatEqn{nd, κ, A}},
+=======
     ::AdvectionDiffusion{1, dim, HeatEqn{nd, κ, A}},
+>>>>>>> origin/master
```
Comment (charleskawczynski) [inline test/Numerics/DGMethods/advection_diffusion/pseudo1D_heat_eqn.jl:62]: Git artifact
